tensorflow_probability/python/experimental/bijectors/highway_flow.py

```diff
@@ -0,0 +1,91 @@
+import numpy as np
+import tensorflow as tf
+
+import tensorflow_probability as tfp
+
+tfd = tfp.distributions
+tfb = tfp.bijectors
+tfe = tf.experimental
+
+'''
+This initial implementation is based on the PyTorch implementation at
+https://github.com/LucaAmbrogioni/CascadingFlow/blob/main/modules/networks.py
+'''
+
+
+class TriResNet(tfb.Bijector):
+
+    # set activation=False for last block
+    def __init__(self, width, activation=True, validate_args=False, name='tri_res_net'):
+        super(TriResNet, self).__init__(
+            validate_args=validate_args,
+            forward_min_event_ndims=0,
+            name=name)
+
+        self.width = width
+        self.W = tf.Variable(np.random.normal(0, 0.01, (self.width, self.width)), trainable=True, dtype=tf.float32)
```

Comment:
In TFP classes we generally assume that `__init__` is a lightweight method that has no 'graph side effects', meaning that it doesn't create any new Tensors or tf.Variables. This is nice because it means it's always cheap to construct an instance, but it does put some constraints on designs.

Another TFP convention (#6 in the style guide: https://github.com/tensorflow/probability/blob/master/STYLE_GUIDE.md) is that we prefer to use descriptive, plain-English names in place of mathematical notation, even if doing so makes the code painfully verbose. For example, we'd prefer `weights` rather than `W`, `bias` rather than `b`, `diagonal` (or similar) rather than `d`, `residual_fraction` (or similar) rather than `lambda`, etc. This makes it easier for a future reader or maintainer of this code (who might not have full mathematical context) to glance at a line of code and at least have some idea of what's going on. (if needed, we can also add a comment to describe the math specifically).

Here I'd suggest that the bijector `__init__` could have a signature along the lines of `__init__(self, weights, bias, diagonal_elements, unconstrained_residual_fractions, activation_fn=tf.nn.sigmoid, validate_args=False, name='tri_res_net')`, where we pass in the variables explicitly using descriptive names (it'd also be nice to accept arbitrary activation functions or `None`, rather than just sigmoid, though that's not essential for a first pass).

Reply:
To clarify re taking variables as `__init__` arguments: of course variables need to be created somewhere, but we'd probably move that to a wrapper function, which could have a signature close to what you have here. Something along the lines of:

```
def build_highway_flow_layer(width, activation):
  return TriResNet(
    weights=tf.Variable(np.random.normal(0, 0.01, (width, width)).astype(tf.float32)),
    bias=tf.Variable(np.random.normal(0, 0.01, (width,)).astype(tf.float32)),
    ...,
    activation=activation)
```